tests/unit/features/overlay_partitions/test_executor_part_handler.py

```diff
@@ -0,0 +1,560 @@
+        assert Path(f"{step_dir}/dir1/baz").exists()
+        assert Path(f"overlay/{step_dir}_overlay").exists()
+
+        self._p1_handler.clean_step(step)
+        assert Path(f"{step_dir}/dir1/.wh.foo").exists() is False
+        assert Path(f"{step_dir}/dir2").exists() is False
+        assert Path(f"{step_dir}/bar").exists() is False
+        assert Path(f"partitions/mypart/{step_dir}/bar").exists() is False
+        assert Path(f"{step_dir}/dir1/baz").exists() is False
+        assert Path(f"overlay/{step_dir}_overlay").exists() is False
+
+    @pytest.mark.parametrize(
+        ("step", "step_dir"), [(Step.STAGE, "stage"), (Step.PRIME, "prime")]
+    )
+    def test_clean_stage_overlay_multiple_parts_with_partitions(self, step, step_dir):
+        _run_step_migration(self._p1_handler, step)
+        assert Path(f"{step_dir}/dir1/foo").exists()
+        assert Path(f"{step_dir}/bar").exists() is False
+        assert Path(f"partitions/mypart/{step_dir}/bar").exists()
+        assert Path(f"{step_dir}/dir1/baz").exists()
+        assert Path(f"overlay/{step_dir}_overlay").exists()
+
+        _run_step_migration(self._p2_handler, step)
+
+        self._p1_handler.clean_step(step)
+        assert Path(f"{step_dir}/dir1/foo").exists()
+        assert Path(f"{step_dir}/bar").exists() is False
+        assert Path(f"partitions/mypart/{step_dir}/bar").exists()
+        assert Path(f"{step_dir}/dir1/baz").exists()
+        assert Path(f"overlay/{step_dir}_overlay").exists()
+
+        self._p2_handler.clean_step(step)
+        assert Path(f"{step_dir}/dir1/foo").exists() is False
+        assert Path(f"{step_dir}/bar").exists() is False
+        assert Path(f"partitions/mypart/{step_dir}/bar").exists() is False
+        assert Path(f"{step_dir}/dir1/baz").exists() is False
+        assert Path(f"overlay/{step_dir}_overlay").exists() is False
+
+    @pytest.mark.parametrize(
+        ("step", "step_dir"), [(Step.STAGE, "stage"), (Step.PRIME, "prime")]
+    )
+    def test_clean_overlay_shared_file_with_partitions(self, mocker, step, step_dir):
+        Path("parts/p1/layer/file1").write_text("content")
+        Path("parts/p3/install/file1").write_text("content")
+
+        _run_step_migration(self._p2_handler, step)
+        _run_step_migration(self._p3_handler, step)
+        assert Path(f"{step_dir}/file1").exists()
+        assert Path(f"{step_dir}/bar").exists() is False
+        assert Path(f"partitions/mypart/{step_dir}/bar").exists()
+
+        # clean overlay data
+        self._p1_handler.clean_step(step)
+        assert Path(f"{step_dir}/bar").exists() is False
+        assert Path(f"partitions/mypart/{step_dir}/bar").exists() is False
+        assert Path(f"{step_dir}/file1").exists()  # file1 remains (also belongs to p3)
+
+        # clean part data
+        self._p3_handler.clean_step(step)
+        assert Path(f"{step_dir}/file1").exists() is False
+
+
+def _run_step_migration(handler: PartHandler, step: Step) -> None:
```

Comment:
This function is now defined twice with the same implementation. Maybe it can be moved to a common place.

Reply:
We can do it in a separate PR.